Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift

```diff
@@ -48,16 +49,49 @@ public extension SwiftSyntaxRule {
         }
 
         return visitor
-            .walk(file: file, handler: \.violationPositions)
+            .walk(file: file, handler: \.violations)
             .sorted()
-            .map { makeViolation(file: file, position: $0) }
+            .map { makeViolation(file: file, violation: $0) }
+    }
+}
+
+/// A violation produced by `ViolationsSyntaxVisitor`s.
+public struct ReasonedRuleViolation: Comparable {
+    /// The violation's position.
+    public let position: AbsolutePosition
+    /// A specific reason for the violation.
+    public let reason: String?
+
+    /// Creates a `ReasonedRuleViolation`.
+    ///
+    /// - parameter position: The violations position in the analyzed source file.
+    /// - parameter reason: The reason for the violation if different from the rule's description.
+    public init(position: AbsolutePosition, reason: String? = nil) {
+        self.position = position
+        self.reason = reason
+    }
+
+    public static func < (lhs: Self, rhs: Self) -> Bool {
+        lhs.position < rhs.position
+    }
+}
+
+/// Extension for arrays of `ReasonedRuleViolation`s that provides the automatic conversion of
+/// `AbsolutePosition`s into `ReasonedRuleViolation`s (without a specific reason).
+extension Array where Element == ReasonedRuleViolation {
+    mutating func append(_ position: AbsolutePosition) {
+        append(ReasonedRuleViolation(position: position))
+    }
+
+    mutating func append(contentsOf positions: [AbsolutePosition]) {
+        positions.forEach { append($0) }
```

Comment:
My gut feeling is that `append(contentsOf: positions.map {  ReasonedRuleViolation(position: $0) })` would be faster